### PR TITLE
[WIP] expand CI coverage

### PR DESCRIPTION
**Pull request by trojikman**

### CODE REVIEW

File: auth_quick_master/controllers/main.py [hunk 1]
```diff
@@ -15,11 +15,11 @@ class AuthQuickMaster(http.Controller):
     @http.route('/auth_quick_master/get-token', type="http", auth='user')
     def get_token(self, build, build_user_id, build_login, build_url):
         _logger.debug('Request for token: build reference = %s, build_user_id = %s, build_login = %s, build_url = %s', build, build_user_id, build_login, build_url)
-        token_obj = request.env['auth_quick_master.token'].create({
+        token_obj = request.env['auth_quick_master.token'].sudo().create({
```
Comment (yelizariev) [inline auth_quick_master/controllers/main.py]: :-1: It allows anyone get access to any build. Revert this and check https://github.com/it-projects-llc/saas-addons/blob/12.0/auth_quick_master/doc/index.rst#configuration

File: saas/models/saas_db.py [hunk 2]
```diff
@@ -41,3 +42,11 @@ def drop_db(self):
     def get_url(self):
         self.ensure_one()
         return self.operator_id.get_db_url(self)
+
+    def action_get_build_access(self):
+        auth_url = '/saas/auth-to-build/' + str(self.id)
+        return {
+            'type': 'ir.actions.act_url',
+            'target': 'self', # or new?
```
Comment (yelizariev) [inline saas/models/saas_db.py]: New browser tab, please

File: saas/models/saas_template.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ class SAASTemplate(models.Model):
         help='Python code to be executed once db is created and modules are installed')
     build_post_init = fields.Text(
         'Build Initialization',
-        default=lambda s: s.env['ir.actions.server'].DEFAULT_PYTHON_CODE,
+        default='',
```
Comment (yelizariev) [inline saas/models/saas_template.py]: :pray:

File: saas/models/saas_template.py [hunk 4]
```diff
@@ -189,7 +190,8 @@ def _rpc_auth(self):
     def _convert_to_dict(key_values):
         key_value_dict = {}
         for r in key_values:
-            key_value_dict.update({r.key: r.value})
+            if r.key and r.value:
+                key_value_dict.update({r.key: r.value})
```
Comment (yelizariev) [inline saas/models/saas_template.py:203]: ``value`` might be empty, actually

File: saas/views/saas_auth.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<!--# Copyright 2019 Denis Mudarisov <https://www.it-projects.info/team/trojikman>
+    # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).-->
+<html>
+<head>
+    <meta charset="UTF-8">
+</head>
+<body>
+  <p style="text-align: center; margin-top: 100px">
+    Within seconds you will be redirected to the Odoo IAP recharge page.
```
Comment (yelizariev) [inline saas/views/saas_auth.html]: IAP, seriously?

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: saas/views/saas_auth.html (hunk 5 above)
Comment (trojikman) [reply saas/views/saas_auth.html]: ooops

File: saas/models/saas_template.py (hunk 4 above)
Comment (trojikman) [reply saas/models/saas_template.py:203]: ok, I will check it

File: saas/models/saas_template.py (hunk 3 above)
Comment (trojikman) [reply saas/models/saas_template.py]: Yes, I will redefine DEFAULT_PYTHON_CODE to correct it

File: auth_quick_master/controllers/main.py (hunk 1 above)
Comment (trojikman) [reply auth_quick_master/controllers/main.py]: I get an access error when run it previous way

File: saas/models/saas_db.py (hunk 2 above)
Comment (trojikman) [reply saas/models/saas_db.py]: Ok

File: auth_quick_master/controllers/main.py (hunk 1 above)
Comment (yelizariev) [reply auth_quick_master/controllers/main.py]: You need to figure out why. You cannot add ``sudo()`` everywhere

Comment (trojikman) [reply auth_quick_master/controllers/main.py]: `sudo()` was here before the commit, but it was after `create()`, so I decided that this method is used incorrectly

Comment (trojikman) [reply auth_quick_master/controllers/main.py]: >  It allows anyone get access to any build. Revert this and check https://github.com/it-projects-llc/saas-addons/blob/12.0/auth_quick_master/doc/index.rst#configuration

I don't think so. Auth quick modules use `sudo()` everywhere in controllers but you can't login to build database without logging in to the master database

Comment (yelizariev) [reply auth_quick_master/controllers/main.py]: Ok, tell me how to allow some users to  use this feature and don't allow to others

Comment (trojikman) [reply auth_quick_master/controllers/main.py]: I fixed this piece of code

File: saas/models/saas_template.py [hunk 6]
```diff
@@ -214,3 +214,18 @@ def create_db(self, db_name, key_values):
         self.operator_id.with_delay().build_post_init(build, self.template_id.build_post_init, key_value_dict)
 
         return build
+
+
+class IrActionsServer(models.Model):
+    _name = 'ir.actions.server.custom.default'
+    _inherit = 'ir.actions.server'
+
+    DEFAULT_PYTHON_CODE = """# Available variables:
```
Comment (yelizariev) [inline saas/models/saas_template.py]: :-1: You don't need to create a table just to introduce a constant

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: saas/models/saas_template.py (hunk 6 above)
Comment (trojikman) [reply saas/models/saas_template.py]: ok

**Comment by trojikman** [comment]
> You might need to add ir.rule additionally. That rule will allow read access to user's tokens only. Otherwise, the malicious user can read all tokens and get unauthrorised access

Done

File: auth_quick_master/security/ir_rule_data.xml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 Denis Mudarisov <https://it-projects.info/team/trojikman>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo>
+    <data noupdate="1">
+        <record model="ir.rule" id="auth_quick_master_token_rule">
+            <field name="name">Token user's only</field>
+            <field name="model_id" ref="model_auth_quick_master_token"/>
+            <field name="domain_force">[('user_id','!=', user.id)]</field>
+            <field name="perm_read" eval="False"/>
```
Comment (yelizariev) [inline auth_quick_master/security/ir_rule_data.xml]: wrong

File: auth_quick_master/security/ir_rule_data.xml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 Denis Mudarisov <https://it-projects.info/team/trojikman>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo>
+    <data noupdate="1">
+        <record model="ir.rule" id="auth_quick_master_token_rule">
+            <field name="name">Token user's only</field>
```
Comment (yelizariev) [inline auth_quick_master/security/ir_rule_data.xml]: what

File: auth_quick_master/security/ir_rule_data.xml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+            <field name="model_id" ref="model_auth_quick_master_token"/>
+            <field name="domain_force">[('user_id','!=', user.id)]</field>
```
Comment (yelizariev) [inline auth_quick_master/security/ir_rule_data.xml]: It must be ```=```

File: auth_quick_master/security/ir_rule_data.xml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 Denis Mudarisov <https://it-projects.info/team/trojikman>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo>
+    <data noupdate="1">
```
Comment (yelizariev) [inline auth_quick_master/security/ir_rule_data.xml]: Don't use ``<data>``. You can put ``noupdate`` to ``<odoo>``

File: auth_quick_master/security/ir_rule_data.xml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 Denis Mudarisov <https://it-projects.info/team/trojikman>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo noupdate="1">
+        <record model="ir.rule" id="auth_quick_master_token_rule">
+            <field name="name">User's token only</field>
+            <field name="model_id" ref="model_auth_quick_master_token"/>
+            <field name="domain_force">[('user_id','=', user.id)]</field>
+            <field name="perm_read" eval="True"/>
```
Comment (yelizariev) [inline auth_quick_master/security/ir_rule_data.xml]: It's True by default

Comment (trojikman) [reply auth_quick_master/security/ir_rule_data.xml]: didn't notice this moment

**Pull request merged**